### PR TITLE
Add callbacks and UserState abstraction

### PR DESCRIPTION
**Pull request by mhanne**
It's been a while, but I just noticed there wasn't a PR for this. I think I sent you a message about it, but that is probably lost with githubs private messaging feature...

Anyway, this makes it easy to integrate ffi-otr into an application (see https://github.com/brettstimmerman/jabber-bot/pull/10/commits/95b8f6fa1fdcde314c14bcb555e446ee2c3a946d for example). There are probably some issues with it, but it seems to work well chatting with pidgin.

### CODE REVIEW

File: lib/ffi/otr/callbacks.rb [hunk 1]
```diff
@@ -0,0 +1,116 @@
+module FFI::OTR::Callbacks
+
+  # Return the OTR policy for the given context.
+  def policy opdata, context
```
Comment (postmodern) [inline lib/ffi/otr/callbacks.rb]: Not to split hairs, but could you put parenthesis around method arguments?

File: lib/ffi/otr/callbacks.rb [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    @opts[:policy]
+  end
+
+  # Create a private key for the given accountname/protocol if desired.
+  def create_privkey opdata, account, protocol
+    if @opts[:privkey]
+      puts "Generating private key. This may take a while."
```
Comment (postmodern) [inline lib/ffi/otr/callbacks.rb]: Eh, I would prefer not to have `puts` in library code. Perhaps these methods could yield or maybe invoke some `@callbacks`?

File: lib/ffi/otr/user_state.rb [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  # @option opts [String] :fingerprints (nil) filename where fingerprints are stored
+  # @option opts [Fixnum] :policy (POLICY_DEFAULT) OTR policy
+  # @option opts [Fixnum] :max_message_size (1024) Max message size
+  # If the :privkey option is set, load the key and implement the
+  # {#create_privkey_cb} callback to generate it.
+  # If the :fingerprints options is set, load fingerprints and implement the
+  # {#new_fingerprint_cb} and {#write_fingerprints_cb} callbacks.
+  # You can always override the callbacks, but if you don't specify those files,
+  # you *must* implement them yourself.
+  def initialize account, protocol, opts = DEFAULT_OPTS
```
Comment (postmodern) [inline lib/ffi/otr/user_state.rb]: `opts = DEFAULT_OPTS` is redundant since we call `DEFAULT_OPTS.merge(opts)`?

File: lib/ffi/otr/user_state.rb [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      @ui_ops["#{name}_cb".to_sym] = ->(*a) { method(name).call(*a) }
+    end
+  end
+
+  # You can call any `otrl_*` method without the `otrl_*`-prefix, and the
+  # userstate is passed in as the first argument. For example:
+  #  userstate.privkey_write_fingerprints("fingerprints.otr")
+  # @note This only works for methods that take the userstate as first argument.
+  def method_missing name, *args
+    if meth = method("otrl_#{name}")
```
Comment (postmodern) [inline lib/ffi/otr/user_state.rb]: Probably want `responds_to?` or `methods.include?`. You may also want to define `responds_to_missing?` to compliment `method_missing`.

**Comment by postmodern** [comment]
Amazing work! I only had a few minor issues with some of the code.

**Comment by mhanne** [comment]
Added a `:debug` option and writing the messages to `$stderr` only when it is set.
They are just default implementations of the callbacks after all, to be overridden by the user as needed.
If they don't want the `Generating private key` message and notify their users in some other way, it's easier to just implement `create_privkey`, rather than having separate `creating_privkey` and `created_privkey` callbacks.

If that's OK with you, I would squash these changes into a single commit.

**Comment by postmodern** [comment]
@mhanne thank you! Squash away and I will merge this.

**Comment by mhanne** [comment]
Made it work with the current version of OTR, since 3.2.1 was pretty old.

**Comment by postmodern** [comment]
Huge thank you! If you are interested in maintaining ffi-otr, I can add you as a contributor. At the least, I will add you to the list of authors.

**Comment by mhanne** [comment]
Sure, I can take care of issues with it.

Still want to get SMP authentication to work, and maybe we can use ffi-gen to generate most of the FFI code automatically. (It works already for surprisingly complex libraries, but OTR seems to be kinda special in how it does its callbacks)

**Pull request merged**